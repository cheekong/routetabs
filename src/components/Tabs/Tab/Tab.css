[role="tab"],
[role="tab"]:focus,
[role="tab"]:hover {
  position: relative;
  margin-top: 5px;
  overflow: visible;
  font-size: 20px;
  border: 0;
  cursor: pointer;
}

[role="tab"][aria-selected="true"] {
  padding: 10px 30px;
  background: darkcyan;
  color: white;
  border: 3px solid darkcyan;
}

[role="tab"] {
  display: inline-block;
  padding: 10px 30px;
  border: 3px solid white;
  background: white;
}

[role="tab"]:hover,
[role="tab"]:focus,
[role="tab"]:active {
  border: 3px solid black;
}
