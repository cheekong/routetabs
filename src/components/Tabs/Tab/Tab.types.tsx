import { ReactNode } from "react";

export interface TabProps {
  id: string;
  children: ReactNode;
  onClick: () => void;
  ariaControls: string;
  tabIndex?: number;
}
