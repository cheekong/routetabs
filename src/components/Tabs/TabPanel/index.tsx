export * from "./TabPanel";
export * from "./TabPanel.types";
