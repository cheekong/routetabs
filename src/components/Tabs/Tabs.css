.tab-group {
  width: 800px;
  border: 1px solid grey;
  border-radius: 4px;
  padding: 10px;
}
