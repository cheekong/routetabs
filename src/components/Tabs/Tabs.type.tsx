import { ReactNode } from "react";

export interface TabsProps {
  id: string;
  defaultTab: string;
  children: ReactNode;
}

export interface TabsContextArgs {
  selectedTab: string;
  setSelectedTab: (tab: string) => void;
}
