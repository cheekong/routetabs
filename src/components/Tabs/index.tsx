export * from "./Tabs";
export * from "./Tabs.type";
export * from "./TabList";
export * from "./Tab";
export * from "./TabPanel";
