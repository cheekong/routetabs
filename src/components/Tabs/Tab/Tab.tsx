import React, { FC, useContext } from "react";
import { isSelected } from "../../../utils";

import { TabsContext } from "..";
import "./Tab.css";
import { TabProps } from "./Tab.types";
import { useNavigate } from "react-router-dom";

export const Tab: FC<TabProps> = ({
  id,
  children,
  ariaControls,
  tabIndex,
  path,
}) => {
  const { selectedTab, setSelectedTab } = useContext(TabsContext);
  const isTabSelected = isSelected(selectedTab, id);
  const navigate = useNavigate();

  return (
    <button
      id={id}
      data-testid={id}
      type="button"
      role="tab"
      tabIndex={
        typeof tabIndex !== "undefined" ? tabIndex : isTabSelected ? 0 : -1
      }
      aria-selected={isTabSelected ? "true" : "false"}
      aria-controls={ariaControls}
      onClick={() => {
        setSelectedTab(path);
        navigate(path);
      }}
    >
      {children}
    </button>
  );
};
