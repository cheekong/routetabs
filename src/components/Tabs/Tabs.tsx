import React, { FC, useEffect, useState } from "react";
import { TabsContextArgs, TabsProps } from ".";
import "./Tabs.css";

export const TabsContext = React.createContext<TabsContextArgs>({
  selectedTab: "/",
  setSelectedTab: (tab: string) => {},
});

export const Tabs: FC<TabsProps> = ({ id, children, defaultTab }) => {
  const [selectedTab, setSelectedTab] = useState(defaultTab);

  useEffect(() => {
    setSelectedTab(defaultTab);
  }, [defaultTab]);

  const contextValue = React.useMemo(
    () => ({
      setSelectedTab,
      selectedTab,
    }),
    [setSelectedTab, selectedTab]
  );

  return (
    <TabsContext.Provider value={contextValue}>
      <div id="carbTab" className="tabs">
        {children}
      </div>
    </TabsContext.Provider>
  );
};
