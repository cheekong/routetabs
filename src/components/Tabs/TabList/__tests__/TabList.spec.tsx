import { TabList, TabListProps } from "../";
import { render, screen } from "@testing-library/react";
import { TabsContext } from "../../Tabs";
import { Tab } from "../../Tab/Tab";
import userEvent from "@testing-library/user-event";

interface RenderProps extends TabListProps {
  selectedTab: string;
  setSelectedTab: () => void;
}

const renderComponent = ({
  selectedTab,
  setSelectedTab,
  children,
  ...tabListProps
}: RenderProps) => {
  return render(
    <TabsContext.Provider value={{ selectedTab, setSelectedTab }}>
      <TabList {...tabListProps}>{children}</TabList>
    </TabsContext.Provider>
  );
};

describe("TabList", () => {
  it("renders tab buttons correctly, have expected arial attributes and able to use keyboard to navigate tabs", () => {
    const { container } = renderComponent({
      ariaLabel: "testTabList",
      ariaLabelLedby: "testLedBy",
      children: (
        <>
          <Tab id="potato" ariaControls="potatoTabPanel" onClick={jest.fn()}>
            Potato
          </Tab>
          <Tab id="noodle" ariaControls="noodleTabPanel" onClick={jest.fn()}>
            Noodle
          </Tab>
        </>
      ),
      selectedTab: "potato",
      setSelectedTab: () => null,
    });

    const tabListElement = screen.getByTestId("testTabList");
    const tabOneElement = screen.getByTestId("potato");
    const tabTwoElement = screen.getByTestId("noodle");

    expect(container).toMatchSnapshot();
    expect(tabOneElement).toBeTruthy();
    expect(tabTwoElement).toBeTruthy();
    expect(tabListElement).toHaveAttribute("role", "tablist");
    expect(tabListElement).toHaveAttribute("aria-label", "testTabList");
    expect(tabListElement).toHaveAttribute("aria-labelledby", "testLedBy");
    expect(tabListElement).toHaveAttribute("role", "tablist");

    // test for accessbility
    userEvent.tab();
    expect(tabOneElement).toHaveFocus();

    userEvent.keyboard("[ArrowRight]");
    expect(tabTwoElement).toHaveFocus();

    userEvent.keyboard("[ArrowLeft]");
    expect(tabOneElement).toHaveFocus();

    userEvent.keyboard("[Home]");
    expect(tabOneElement).toHaveFocus();

    userEvent.keyboard("[End]");
    expect(tabTwoElement).toHaveFocus();
  });
});
