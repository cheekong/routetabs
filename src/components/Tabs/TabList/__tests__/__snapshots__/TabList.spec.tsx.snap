// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList renders tab buttons correctly, have expected arial attributes and able to use keyboard to navigate tabs 1`] = `
<div>
  <div
    aria-label="testTabList"
    aria-labelledby="testLedBy"
    data-testid="testTabList"
    role="tablist"
  >
    <button
      aria-controls="potatoTabPanel"
      aria-selected="true"
      data-testid="potato"
      id="potato"
      role="tab"
      tabindex="0"
      type="button"
    >
      Potato
    </button>
    <button
      aria-controls="noodleTabPanel"
      aria-selected="false"
      data-testid="noodle"
      id="noodle"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Noodle
    </button>
  </div>
</div>
`;
